### PR TITLE
HWKAPM-511 Fix issue with business transaction name not being propaga…

### PR DESCRIPTION
**Pull request by objectiser**
…ted to subsequent simulated service's trace information

### CODE REVIEW

**Comment by objectiser** [comment]
Review notes:
* This application is a client simulator, indirectly producing trace information for the execution of different service invocation patterns.
* It is also possible to define business transaction configurations on the server, to classify the simulated traces as belonging to specific business transactions.
* However when this is done (i.e. the top level service/a URI tagged as being "MyBTxn"), when the trace data is received on the server, only the trace fragment for the initial service/a invocation had the business txn name - not the subsequent called service/b or service/c services.
* So this fix sorts out that problem - however to create a suitable unit test, I needed to make some additional changes to the code.

**Comment by objectiser** [comment]
@lucasponce Another one, if you have some time. I'm away tomorrow so not urgent. If you have less time now, let me know and I'll find someone else. Thanks.

File: performance/server/src/main/java/org/hawkular/apm/performance/server/DefaultServiceRegistry.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        synchronized (stack) {
+            if (stack.isEmpty()) {
+                // Allocate new instance
+                return newServiceInstance(name);
+            } else {
+                Service service = stack.pop();
+
+                // Check service hasn't expired
+                if (System.currentTimeMillis() - service.getLastUsed() > 10000) {
```
Comment (lucasponce) [inline performance/server/src/main/java/org/hawkular/apm/performance/server/DefaultServiceRegistry.java]: Minor one.
10000 shouldn't be a constant that can be parametrized ?

File: performance/server/src/main/java/org/hawkular/apm/performance/server/Service.java [hunk 2]
```diff
@@ -35,7 +34,7 @@
 
     private String name;
 
-    private Map<String, String> calledServices = new HashMap<String, String>();
+    private Map<String, String> calledServices;
```
Comment (lucasponce) [inline performance/server/src/main/java/org/hawkular/apm/performance/server/Service.java]: Is this initialized in a different point ?
I can't see it on the patch and I see a reference of this one in some code below.

File: performance/server/src/test/java/org/hawkular/apm/performance/server/ServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * @author gbrown
+ */
+public class ServiceTest {
+
+    /**  */
+    private static final String TXN_NAME = "TestBTxnName";
+
+    @Test
+    public void testBusinessTxnNamePropagated() {
+        TestTracePublisher publisher = (TestTracePublisher) ServiceResolver.getSingletonService(TracePublisher.class);
```
Comment (lucasponce) [inline performance/server/src/test/java/org/hawkular/apm/performance/server/ServiceTest.java]: Minor one.
Do you need to access to the impl class instead just the TracePublisher interface ?
Just if that could create some dependency on the Impl somewhere.
But anyway, I am not fully familiar with the code, perhaps this comment has not sense at al.

**Comment by lucasponce** [comment]
Looks good to me.
I've made a couple of minor comments in case it makes sense.

File: performance/server/src/main/java/org/hawkular/apm/performance/server/DefaultServiceRegistry.java (hunk 1 above)
Comment (objectiser) [inline performance/server/src/main/java/org/hawkular/apm/performance/server/DefaultServiceRegistry.java]: Agree - I'll sort that out.

File: performance/server/src/main/java/org/hawkular/apm/performance/server/Service.java (hunk 2 above)
Comment (objectiser) [inline performance/server/src/main/java/org/hawkular/apm/performance/server/Service.java]: Yes it is supplied in the constructor, so creating it here was obsolete, so decided to remove it.

File: performance/server/src/test/java/org/hawkular/apm/performance/server/ServiceTest.java (hunk 3 above)
Comment (objectiser) [inline performance/server/src/test/java/org/hawkular/apm/performance/server/ServiceTest.java]: Understand, but in this case the TracePublisher impl is hidden in the collector mechanism, so not accessible - so by using a known 'test' implementation I'm able to inspect what Traces have been created.

**Comment by objectiser** [comment]
@lucasponce Thanks!

**Pull request merged**